zkevm-circuits/src/tx_circuit.rs

```diff
@@ -0,0 +1,542 @@
+    let r_be = r.to_be_bytes();
+    let s_be = s.to_be_bytes();
+    let mut r = libsecp256k1::curve::Scalar::from_int(0);
+    let r_overflow: bool = r.set_b32(&r_be).into();
+    let mut s = libsecp256k1::curve::Scalar::from_int(0);
+    let s_overflow: bool = s.set_b32(&s_be).into();
+    if r_overflow || s_overflow {
+        error!("Overflow on 'r' or 's' signature values");
+        return Err(Error::Synthesis);
+    }
+    let signature = libsecp256k1::Signature { r, s };
+    let msg_hash = libsecp256k1::Message::parse_slice(msg_hash.as_slice()).map_err(|e| {
+        error!("Message hash parsing from slice failed: {:?}", e);
+        Error::Synthesis
+    })?;
+    let recovery_id = libsecp256k1::RecoveryId::parse(v).map_err(|e| {
+        error!("secp256k1::RecoveriId::parse error: {:?}", e);
+        Error::Synthesis
+    })?;
+    let pk = libsecp256k1::recover(&msg_hash, &signature, &recovery_id).map_err(|e| {
+        error!("Public key recovery failed: {:?}", e);
+        Error::Synthesis
+    })?;
+    let pk_be = pk.serialize();
+    let mut pk_le = [0u8; 64];
+    pk_le.copy_from_slice(&pk_be[1..]);
+    pk_le[..32].reverse();
+    pk_le[32..].reverse();
+    let mut pk_bytes = secp256k1::Serialized::default();
+    pk_bytes.as_mut().copy_from_slice(&pk_le[..]);
+    let pk = Secp256k1Affine::from_bytes(&pk_bytes);
+    ct_option_ok_or(pk, Error::Synthesis).map_err(|e| {
+        error!("Invalid public key little endian bytes");
+        e
+    })
+}
+
+fn biguint_to_32bytes_le(v: BigUint) -> [u8; 32] {
+    let mut res = [0u8; 32];
+    let v_le = v.to_bytes_le();
+    res[..v_le.len()].copy_from_slice(&v_le);
+    res
+}
+
+fn ct_option_ok_or<T, E>(v: CtOption<T>, err: E) -> Result<T, E> {
+    Option::<T>::from(v).ok_or(err)
+}
+
+fn tx_to_sign_data(tx: &Transaction, chain_id: u64) -> Result<SignData, Error> {
+    let sig_r_le = tx.r.to_le_bytes();
+    let sig_s_le = tx.s.to_le_bytes();
+    let sig_r =
+        ct_option_ok_or(secp256k1::Fq::from_repr(sig_r_le), Error::Synthesis).map_err(|e| {
+            error!("Invalid 'r' signature value");
+            e
+        })?;
+    let sig_s =
+        ct_option_ok_or(secp256k1::Fq::from_repr(sig_s_le), Error::Synthesis).map_err(|e| {
+            error!("Invalid 's' signature value");
+            e
+        })?;
+    // msg = rlp([nonce, gasPrice, gas, to, value, data, sig_v, r, s])
+    let mut stream = RlpStream::new_list(9);
+    stream
+        .append(&tx.nonce)
+        .append(&tx.gas_price)
+        .append(&tx.gas)
+        .append(&tx.to.unwrap_or_else(Address::zero))
+        .append(&tx.value)
+        .append(&tx.data.0)
+        .append(&chain_id)
+        .append(&0u32)
+        .append(&0u32);
+    let msg = stream.out();
+    let msg_hash: [u8; 32] = Keccak256::digest(&msg)
+        .as_slice()
+        .to_vec()
+        .try_into()
+        .expect("hash length isn't 32 bytes");
+    let v = (tx.v - 35 - chain_id * 2) as u8;
```

Comment:
This comes from https://github.com/appliedzkp/zkevm-specs/blob/master/specs/transactions-proof.md#legacy-type
An ethereum signature contains three integers (v, r, s).  But actually only (r, s) is the ECDSA signature.  `v` is a parameter that merges the a bit value required to recover the public key from a msg_hash and ECDSA signature, and the chain_id like so:
`sig_v = {0,1} + CHAIN_ID * 2 + 35`

So in the code, `sig_v = tx.v`, `{0,1} = v`.  And we get back this bit.  If the `chain_id` is the correct one used for the ethereum signature, then `v` should be `{0, 1}` and fit into `u8`.  If not, there may be an underflow, an overflow, and even after that the `recover_pk` function will return an error.  I can add a check to make sure there are no underflows or overflows, and return an error.  Does that sound good?

Reply:
So. I've thought about this a bit more. 
I guess that it shouldn't be an issue to let the underflow pass and continue computing. 
The problem is that we know we will generate a proof that is not valid. And so, I'm not sure till which point we want to avoid or not erroring here.

Also, I know that for our use-case, we don't really care as basically, the `chain_id = 1` and we could hardcode that. But since we did not hardcode in the circuit Opcode impl for ChainId, I think we shouldn't here neither.

Leave it your to you, just wanted to raise a point and understand the situation and rationale :smile:

Reply:
> So. I've thought about this a bit more. I guess that it shouldn't be an issue to let the underflow pass and continue computing. The problem is that we know we will generate a proof that is not valid. And so, I'm not sure till which point we want to avoid or not erroring here.

Oh but there will be an error reported.  If the chain_id is not the correct one, a proof will not be generated because the function `recover_pk` will return an error.

> Also, I know that for our use-case, we don't really care as basically, the `chain_id = 1` and we could hardcode that. But since we did not hardcode in the circuit Opcode impl for ChainId, I think we shouldn't here neither.

I think we care!  `chain_id = 1` is only for proofs of validity of mainnet, but for any other use case (like a zkrollup) we must support any chain_id.